demo/api/src/products/product-persist.pipe.ts

```diff
@@ -0,0 +1,129 @@
+import { LoggerService } from "@nestjs/common";
+import { ImporterPipe, PipeMetadata } from "@src/importer/pipes/importer-pipe.type";
+import { Transform, TransformCallback } from "stream";
+import { v4 } from "uuid";
+
+import { Product } from "./entities/product.entity";
+import { ProductCategory } from "./entities/product-category.entity";
+import { ProductColor } from "./entities/product-color.entity";
+import { ProductImporterInput } from "./product-importer.input";
+
+type ProductData = Omit<ProductImporterInput, "image"> & {
+    id: string;
+    colors: string;
+    category: Reference<ProductCategory>;
+    image: BlockDataInterface;
+};
+
+export class ProductPersistPipe implements ImporterPipe {
+    constructor(private readonly em: EntityManager<IDatabaseDriver<Connection>>) {}
+
+    getPipe(logger: LoggerService) {
+        return new ProductPersist(this.em, logger);
+    }
+}
+
+class ProductPersist extends Transform {
+    private persistedEntitiesAmount = 0;
+
+    constructor(private readonly em: EntityManager<IDatabaseDriver<Connection>>, private readonly logger: LoggerService) {
+        super({ writableObjectMode: true, objectMode: true });
+    }
+
+    getPersistedEntitiesAmount() {
+        return this.persistedEntitiesAmount;
+    }
+
+    async _transform(inputDataAndMetadata: { data: ProductData; metadata: PipeMetadata }, encoding: BufferEncoding, callback: TransformCallback) {
+        try {
+            const { data } = inputDataAndMetadata;
+            const { colors: colorString, category, colors: colorCollection, ...productData } = data;
+
+            const updateQuery: FilterQuery<object> | undefined = { slug: data.slug };
+            const record: Product | null = updateQuery ? await this.em.findOne(Product, updateQuery, { fields: ["id"] }) : null;
+
+            productData.image = DamImageBlock.blockInputFactory({
+                activeType: "pixelImage",
+                attachedBlocks: [
+                    {
+                        props: {},
+                        type: "pixelImage",
+                    },
+                ],
+            }).transformToBlockData();
+
+            // TODO: use upsert
+            // const product = await this.em.upsert(Product, productInput, {
+            //     onConflictFields: ["slug"],
+            //     onConflictAction: "merge",
+            //     onConflictExcludeFields: ["id", "createdAt"],
+            // });
+            if (record) {
+                await this.logger.log(`Existing record for ${JSON.stringify(updateQuery)} with id ${record.id} found, updating...`);
+                // keeping the id from the input data (in the DataTransformer plainToInstance always creates a new Entity instance, which is populated with the new id)
+                data.id = record.id;
+                await this.em.nativeUpdate(Product.name, { id: record.id }, { ...productData, category, id: data.id, updatedAt: new Date() });
```

Comment:
Why do you call nativeUpdate and not update?

Reply:
@paha77 Do you remember why we use nativeUpdate in the other project?

Reply:
Because of performance reasons.

Reply:
Should we add a comment for this?